### PR TITLE
[AN-5241] images restart downloading when switching network

### PR DESCRIPTION
**Pull request by makingthematrix**

### CODE REVIEW

File: zmessaging/src/main/scala/com/waz/ui/UiModule.scala [hunk 1]
```diff
@@ -138,6 +138,7 @@ class UiModule(val accounts: AccountsService) extends UiEventContext with ZMessa
   lazy val assets = new UiCache[AssetId, Asset](10)(this)
 
   lazy val globalImageLoader = global.imageLoader
+  lazy val globalNetwork = global.network
```
Comment (deanrobertcook) [inline zmessaging/src/main/scala/com/waz/ui/UiModule.scala]: `network` should be enough - there are no non-global network services.

**Comment by deanrobertcook** [comment]
Also - as we discussed on Friday - look at removing the `vars` and only setting the listener on `DownloadFailedWifiException` (or whatever it's called)

File: tests/unit/src/test/scala/com/waz/service/images/BitmapSignalSpec.scala [hunk 2]
```diff
@@ -17,128 +17,192 @@
...
+      (loader.loadBitmap _).expects(expectedAsset, req, *).once().onCall { _ =>
+        val bitmap = mockBitmap(expectedAsset)
+        bmps ! Some(bitmap)
+        CancellableFuture.successful(bitmap)
+      }
+
+      (loader.loadCachedBitmap _).expects(expectedAsset, req).anyNumberOfTimes.onCall { _ =>
+        bmps.currentValue.flatten match {
+          case Some(bitmap) => CancellableFuture.successful(bitmap)
+          case None => CancellableFuture.failed(new IllegalStateException("Trying to return a non-existing cached bitmap"))
```
Comment (deanrobertcook) [inline tests/unit/src/test/scala/com/waz/service/images/BitmapSignalSpec.scala]: if this should fail the test, you can also just call `fail("some failure message")`

File: zmessaging/src/main/scala/com/waz/service/images/BitmapSignal.scala [hunk 3]
```diff
@@ -20,60 +20,63 @@ package com.waz.service.images
...
-        loader.load() flatMap { loader.process(_, signal) }
+    future = loader.loadCached().flatMap {
+      case Some(data) => loader.process(data, signal)
+      case None => loader.load().flatMap { loader.process(_, signal) } // will try with download
     }
     future onFailure {
-      case _: CancelException => // ignore
+      case _: CancelException =>
+      case ex@AssetLoader.DownloadOnWifiOnlyException =>
+        signal publish LoadingFailed(ex)
```
Comment (deanrobertcook) [inline zmessaging/src/main/scala/com/waz/service/images/BitmapSignal.scala]: you can just use `signal ! LoadingFailed(ex)`

**Comment by deanrobertcook** [comment]
merge whenever your happy, and then please create a release with the appropriate fixes :)

File: zmessaging/src/main/scala/com/waz/service/images/BitmapSignal.scala (hunk 3 above)
Comment (makingthematrix) [inline zmessaging/src/main/scala/com/waz/service/images/BitmapSignal.scala]: I can't. BitmapSignal is not a SourceSignal. I can use publish, but the "!" shortcut is not defined. (kinda weird)

**Pull request merged**